File: frontend/src/app/core/services/complaint.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Complaint, ComplaintRequest, ComplaintStatusUpdateRequest } from '../models/complaint.model';
import { PageResponse } from '../models/page-response.model';
import { ComplaintType } from '../models/enums';

export interface ComplaintSearchCriteria {
  customerCode?: string;
  consumerNumber?: string;
  complaintNumber?: string;
  complaintType?: string;
  status?: string;
  fromDate?: string;
  toDate?: string;
}

@Injectable({ providedIn: 'root' })
export class ComplaintService {
  private readonly baseUrl = '/api/complaints';

  constructor(private readonly http: HttpClient) {}

  categories(): Observable<Record<ComplaintType, string[]>> {
    return this.http.get<Record<ComplaintType, string[]>>(`${this.baseUrl}/categories`);
  }

  register(request: ComplaintRequest): Observable<Complaint> {
    return this.http.post<Complaint>(this.baseUrl, request);
  }

  track(complaintNumber: string): Observable<Complaint> {
    return this.http.get<Complaint>(`${this.baseUrl}/track/${complaintNumber}`);
  }

  myHistory(page: number, size: number): Observable<PageResponse<Complaint>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResponse<Complaint>>(`${this.baseUrl}/me`, { params });
  }

  search(criteria: ComplaintSearchCriteria, page: number, size: number): Observable<PageResponse<Complaint>> {
    let params = new HttpParams().set('page', page).set('size', size);
    Object.entries(criteria).forEach(([key, value]) => {
      if (value) params = params.set(key, value);
    });
    return this.http.get<PageResponse<Complaint>>(this.baseUrl, { params });
  }

  updateStatus(id: number, request: ComplaintStatusUpdateRequest): Observable<Complaint> {
    return this.http.put<Complaint>(`${this.baseUrl}/${id}/status`, request);
  }

  export(criteria: ComplaintSearchCriteria, format: 'csv' | 'pdf'): Observable<Blob> {
    let params = new HttpParams().set('format', format);
    Object.entries(criteria).forEach(([key, value]) => {
      if (value) params = params.set(key, value);
    });
    return this.http.get(`${this.baseUrl}/export`, { params, responseType: 'blob' });
  }
}

File: backend/src/main/java/com/electricity/billing/service/ComplaintService.java
package com.electricity.billing.service;

import com.electricity.billing.dto.request.ComplaintRequest;
import com.electricity.billing.dto.request.ComplaintStatusUpdateRequest;
import com.electricity.billing.dto.response.ComplaintResponse;
import com.electricity.billing.dto.response.PageResponse;
import com.electricity.billing.entity.enums.ComplaintStatus;
import com.electricity.billing.entity.enums.ComplaintType;

import java.time.LocalDateTime;

public interface ComplaintService {

    ComplaintResponse register(ComplaintRequest request);

    ComplaintResponse getByComplaintNumber(String complaintNumber);

    PageResponse<ComplaintResponse> myHistory(int page, int size);

    PageResponse<ComplaintResponse> search(String customerCode, String consumerNumber, String complaintNumber,
                                            ComplaintType complaintType, ComplaintStatus status,
                                            LocalDateTime fromDate, LocalDateTime toDate, int page, int size);

    ComplaintResponse updateStatus(Long id, ComplaintStatusUpdateRequest request);

    byte[] exportComplaints(String customerCode, String consumerNumber, ComplaintType complaintType,
                             ComplaintStatus status, String format);
}
